processing/src/main/java/io/druid/query/topn/TopNMapFn.java

```diff
@@ -20,12 +20,146 @@
 import io.druid.segment.Cursor;
+import io.druid.segment.DimensionHandlerUtils;
 import io.druid.segment.DimensionSelector;
+import io.druid.segment.column.ColumnCapabilities;
+import io.druid.segment.column.ValueType;
+import io.druid.segment.data.IndexedInts;
+
+import java.util.Map;
 
 public class TopNMapFn implements Function<Cursor, Result<TopNResultValue>>
 {
+  private static final TopNTypeHelperFactory TYPE_HELPER_FACTORY = new TopNTypeHelperFactory();
+
+  private static class TopNTypeHelperFactory implements QueryTypeHelperFactory<TopNTypeHelper>
+  {
+    @Override
+    public TopNTypeHelper makeQueryTypeHelper(
+        String dimName, ColumnCapabilities capabilities
+    )
+    {
+      ValueType type = capabilities.getType();
+      switch(type) {
+        case STRING:
+          return new StringTopNTypeHelper();
+        default:
+          return null;
+      }
+    }
+  }
+
+  public interface TopNTypeHelper<ValueSelectorType extends ColumnValueSelector> extends QueryTypeHelper
+  {
+    /**
+     * Used by DimExtractionTopNAlgorithm.
+     *
+     * Create an Aggregator[][] using BaseTopNAlgorithm.AggregatorArrayProvider and the given parameters.
+     *
+     * As the Aggregator[][] is used as an integer-based lookup, this method is only applicable for dimension types
+     * that use integer row values.
+     *
+     * A dimension type that does not have integer values should return null.
+     *
+     * @param params Parameters for the TopN query being served
+     * @param query The TopN query being served
+     * @param capabilities Object indicating if dimension values are sorted
+     * @return an Aggregator[][] for integer-valued dimensions, null otherwise
+     */
+    Aggregator[][] getDimExtractionRowSelector(TopNParams params, TopNQuery query, Capabilities capabilities);
+
+
+    /**
+     * Used by DimExtractionTopNAlgorithm.
+     *
+     * Read the current row from a dimension value selector, and for each row value:
+     * 1. Retrieve the Aggregator[] for the row value from rowSelector (fast integer lookup) or from
+     *    aggregatesStore (slower map).
+     *
+     * 2. If the rowSelector and/or aggregatesStore did not have an entry for a particular row value,
+     *    this function should retrieve the current Aggregator[] using BaseTopNAlgorithm.makeAggregators() and the
+     *    provided cursor and query, storing them in rowSelector and aggregatesStore
+     *
+     * 3. Call aggregate() on each of the aggregators.
+     *
+     * If a dimension type doesn't have integer values, it should ignore rowSelector and use the aggregatesStore map only.
+     *
+     * @param selector Dimension value selector
+     * @param rowSelector Integer lookup containing aggregators
+     * @param aggregatesStore Map containing aggregators
+     * @param cursor Cursor for the segment being queried
+     * @param query The TopN query being served.
+     */
+    void dimExtractionScanAndAggregate(
```

Comment:
Maybe put `query` parameter first, as "immutable", and rowSelector/aggregatesStore last, as "output" parameters?